pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java

```diff
@@ -1560,15 +1634,26 @@ public TypeNode getValueTypeNode() {
       return valueTypeNode;
     }
 
-    protected ListingOrMappingTypeCastNode getListingOrMappingTypeCastNode() {
+    protected ListingOrMappingTypeCastNode getListingOrMappingTypeCastNode(
+        VirtualFrame frame, boolean executeInDefinition) {
       if (listingOrMappingTypeCastNode == null) {
         CompilerDirectives.transferToInterpreterAndInvalidate();
-        listingOrMappingTypeCastNode =
-            new ListingOrMappingTypeCastNode(
-                VmLanguage.get(this),
-                getRootNode().getFrameDescriptor(),
-                valueTypeNode,
-                getRootNode().getName());
+        if (executeInDefinition) {
+          listingOrMappingTypeCastNode =
+              new ListingOrMappingTypeCastNode(
+                  VmLanguage.get(this),
+                  getRootNode().getFrameDescriptor(),
+                  valueTypeNode,
+                  getRootNode().getName(),
+                  frame);
+        } else {
+          listingOrMappingTypeCastNode =
+              new ListingOrMappingTypeCastNode(
+                  VmLanguage.get(this),
+                  getRootNode().getFrameDescriptor(),
+                  valueTypeNode,
+                  getRootNode().getName());
```

Comment:
It's this kind of repetition that would almost make me suggest to just make passing a `null` part of the API of `ListingOrMappingTypeCastNode`

Reply:
Improved.